### PR TITLE
Enhancement/328 Sophi Overrides API Integration

### LINKED ISSUE #216: Automation block wants to retrieve result set while updating its settings
### Is your enhancement related to a problem? Please describe.

The automation block starts fetching results while typing the page and widget name. While not strictly a problem, consider adding a delay or otherwise only fetch results after the settings for the block are completed

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #328: Add integration for Sophi Overrides API
### Is your enhancement related to a problem? Please describe.

Integrate the [Sophi Override functions](https://documenter.getpostman.com/view/16875817/TzzHmCyw#2a201cca-c05b-4d98-b9fd-5e96ce1baf54) (add, remove, replace, ban) into the plugin experience taking into account sites using the default Sophi Site Automation block, a custom created block, WP_Query / get_posts() overrides, and any additional concerns for headless setups.

### Approach

- [x] Review requirements / comments here and determine technical approach
- [x] If estimate is less than 40hrs proceed, if over 40hrs then reach out to @jeffpaul for clarification on how best to proceed (e.g., build add/remove functions only)
- [x] Build block editor integration for updated Sophi block as nested block with each nested block having an Override block control (add, remove, replace, ban)
- [x] Include hook(s) for non-block editor-based page curation to have access to similar Override functionality (ensure properly documented to appear on [hookdocs](https://globeandmail.github.io/sophi-for-wordpress/index.html))
- [ ] Test on 10up/VIP test sites, validate w/Sophi team, then release
- [ ] Post-release consider an additional WP Admin screen to display a list of Overrides and ability to cancel individual/bulk existing Overrides

### Describe alternatives you've considered

Maintain Slack/MS Team integration

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by faisal-alvi**
### Description of the Change

* Integrated the [Sophi Override functions](https://documenter.getpostman.com/view/16875817/TzzHmCyw#2a201cca-c05b-4d98-b9fd-5e96ce1baf54) (add, remove, replace, ban) into the existing Site Automation block.
* The existing block will not break after this integration because the final code snippet of the block does not change! (i.e. nothing to worry about the migration of existing block).

Closes #328 & Closes #216

### Alternate Designs

_List view:_

image
[image link](https://www.screencast.com/t/zEMLZpusYI9)

---

_The 4 Override buttons:_

image
[image link](https://www.screencast.com/t/0r60bXIIZo)

---

_Add/Replace Override Popup:_

image
[image link](https://www.screencast.com/t/7cSwkE1pC)

---

_Out/Ban Override Popup:_

image
[image link](https://www.screencast.com/t/crkbQS4H)

---
_Settings page:_

image
[image link](https://www.screencast.com/t/JsZlqyDcG7K)

---

_When posts are received from the Sophi server, but none of them exist in the WordPress site:_

image
[image link](https://www.screencast.com/t/Np9gAvUAtGB)

---

_When posts are received from the Sophi server, but none of them exist in the WordPress site:_

image
[image link](https://www.screencast.com/t/I4nYVhKM6s)

### Verification Process

1. Make sure you have configured proper credential details on the Sophi settings page.
2. Add a Site Automation block to a page.
3. Fill in "Page name" and "Widget name" in the block settings on the right sidebar.
4. See the curated post list, it should appear.
5. Select any individual post item and see it has its own settings (`Blockcontrol`) at the top.
6. Click the gear icon, and a popup should appear.
7. Perform Add/Replace/Remove/Ban actions by filling out the form.
8. For add/replace, the result will appear quickly.
9. But for remove/ban, the Sophi server will take some time to reflect.

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### Changelog Entry

> Added - Override feature to the Sophi Site Automation Block.
> Added - `???` hook to allow for code-based Override feature integrations.
> Changed - Sophi Site Automation Block from a server-side-rendered custom block to a nested block.

### Credits

Props @jeffpaul @fabiankaegy @faisal-alvi

### CODE REVIEW

**Comment by jeffpaul** [comment]
Pinging @fabiankaegy for initial code review / feedback on approach here to help guide any iterations or strong recommendations on how best to tweak approach here for an optimal editorial UX.

**Comment by jeffpaul** [comment]
@faisal-alvi wanted to make sure you saw the [6 Bugs](https://sonarcloud.io/project/issues?id=globeandmail_sophi-for-wordpress&pullRequest=333&resolved=false&types=BUG) that were flagged by @sonarcloud in case those are easily handled in your PR?

**Comment by jeffpaul** [comment]
Some block toolbar updates for the nested blocks:
1.) Update the nested block icon from the pushpin to the list item icon from WP 6.1 (@fabiankaegy can help show/link to that)
2.) Hide the lock icon
3.) Replace the gear icon with four icons that will relate to the four Override options (add, replace, remove, ban) using the following icons for each
add: https://developer.wordpress.org/resource/dashicons/#insert (alternatively `addSubmenu` Gutenberg icon)
replace: https://developer.wordpress.org/resource/dashicons/#update (alternatively `replace` Gutenberg icon)
remove: https://developer.wordpress.org/resource/dashicons/#remove (alternatively `removeSubmenu` Gutenberg icon)
ban: https://developer.wordpress.org/resource/dashicons/#trash (alternatively `trash` Gutenberg icon)
Each modal for those 4 new options will now just be the field(s) needed for that specific request and can remove the initial dropdown selection for add/replace/remove/ban.

**Comment by jeffpaul** [comment]
@faisal-alvi please update the plugin settings page to:
1. rename the `API Authentication settings` to `Override settings`
2. move `Override settings` section below the `Site Automation settings` section
3. move the `Override API URL` item into the `Override settings` section
4. add another section heading of `Advanced settings` that includes the `Query Integration` item

File: includes/blocks/sophi-page-list-item/block.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    "overrideRule": {
+      "type": "string"
+    },
+    "overridePostID": {
+      "type": "integer"
+    },
+    "overrideExpiry": {
+      "type": "integer"
+    }
+  },
```
Comment (fabiankaegy) [inline includes/blocks/sophi-page-list-item/block.json:44]: @faisal-alvi In order to hide the lock icon for the block you can add 

```json
{
    "supports": {
        "lock": false
    }
}
```

With that option, the lock still happens but the UI for it doesn't get shown. 

https://gutenberg.10up.com/reference/Blocks/block-supports/#lock

File: includes/blocks/sophi-page-list-item/edit.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	const overridePopover = showPopup && (
+		<Popover
+			className="override-map-popover"
+			position="bottom center"
+			focusOnMount="firstElement"
+			key="override-popover"
+			expandOnMobile
+			headerTitle={__('Override', 'Override')}
+			onFocusOutside={() => setShowPopup(false)}
+		>
```
Comment (fabiankaegy) [inline includes/blocks/sophi-page-list-item/edit.js:120]: I would suggest to also add the `anchorRef` prop here and provide the `ref` of the toolbar button.


```js
import { useRef } from '@wordpress/element';
import { Popover, ToolbarButton } from '@wordpress/components';

const BlockEdit = () => {

    const toolbarButtonRef = useRef();

    return (
        <>
            <ToolbarButton ref={toolbarButtonRef} />
            <Popover anchorRef={toolbarButtonRef} />
        </>
    );
}
```

_Note: This is simplified code that doesn't work. it is only meant to show how the `anchorRef` prop works._

This will fix any positioning issues you are having with the popover and will make sure that it gets attached to the toolbar button :)

File: includes/blocks/sophi-page-list-item/index.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+registerBlockType(block.name, {
+	title: __('Sophi Recommended Post', 'sophi-wp'),
+	description: __(
+		'This is the recommended post from Sophi Site Automation for the respective page and widget combination for the parent Site Automation block.',
+		'sophi-wp',
+	),
+	parent: ['sophi/site-automation-block'],
+	edit,
+	save,
+	icon: { src: 'admin-post' },
```
Comment (fabiankaegy) [inline includes/blocks/sophi-page-list-item/index.js]: Lets use the icon from the Core List Item block here:

https://github.com/WordPress/gutenberg/blob/ba01e5d7851d6211d020687b5ae927f47ea4683e/packages/block-library/src/list-item/index.js#L4

_Note: make sure the `@wordpress/icons` npm package is actually installed in the plugin. This one is not bundled with WordPress Core._

**Comment by faisal-alvi** [comment]
Please note the PHP Compat test for PHP 8.0 is failing since https://github.com/globeandmail/sophi-for-wordpress/commit/4b0703aed714289eeeecbab43ff26e6c8802e347. That should be handled seprately. cc: @Sidsector9

File: includes/blocks/site-automation-block/register.php [hunk 4]
```diff
@@ -24,15 +24,22 @@ function register() {
...
  * @param array  $block      Block information such as context.
  *
- * @return string The rendered block markup.
+ * @return string|\WP_Post[] The rendered block markup OR WP_Post object to be returned to the REST callback.
  */
 function render_block_callback( $attributes, $content, $block ) {
+	$is_gb_editor = \defined( 'REST_REQUEST' ) && REST_REQUEST && ! empty( $_REQUEST['context'] ) && 'edit' === $_REQUEST['context']; // phpcs:ignore WordPress.Security.NonceVerification
+
+	// Render only on the front end.
+	if( $is_gb_editor && 'via_rest' !== $content  ) {
```
Comment (dkotter) [inline includes/blocks/site-automation-block/register.php]:
```suggestion
	if ( $is_gb_editor && 'via_rest' !== $content ) {
```

File: includes/blocks/sophi-page-list-item/index.css [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Example Block
+ * This file is for CSS admin overrides only.
+ * Use theme files for standard CSS work.
+ */
```
Comment (dkotter) [inline includes/blocks/sophi-page-list-item/index.css]: Not sure this comment makes total sense in this context, or is needed

File: includes/classes/SiteAutomation/Auth.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	 * @return string|\WP_Error
+	 */
+	public function refresh_access_token() {
+		$response = $this->request_access_token();
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		set_transient( 'sophi_site_automation_access_token', $response['access_token'], $response['expires_in'] );
```
Comment (dkotter) [inline includes/classes/SiteAutomation/Auth.php]: Do we need to consider encrypting this token before it's stored? I know a lot of times these access tokens aren't encrypted, but where possible, it is best practice to encrypt those in case of a database breach. But I guess need to decide if that's worth the effort here

File: includes/classes/SiteAutomation/Auth.php [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		$response = wp_remote_retrieve_body( $result );
+
+		return json_decode( $response, true );
+	}
+
+	/**
+	 * Set the environment to be used.
+	 *
+	 * @since 1.3.0
+	 * @param string $environment Could be 'prod', 'stg' or 'dev'
```
Comment (dkotter) [inline includes/classes/SiteAutomation/Auth.php]:
```suggestion
	 * @param string $environment Could be 'prod', 'stg' or 'dev'.
```

File: includes/classes/SiteAutomation/EndPoints.php [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+		// POST /sophi/v1/update-posts route.
+		register_rest_route(
+			self::SOPHI_NAMESPACE,
+			'/site-automation-override',
+			array(
+				array(
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => array( $this, 'site_automation_override' ),
+					'permission_callback' => '__return_true',
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: So it seems we check if a user is logged-in within our normal `callback`. Wondering if that would be bette within this `permission_callback`, rather than just returning `true` here?

File: includes/classes/SiteAutomation/EndPoints.php [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+
+		/**
+		 * Filters the get sophi posts query params.
+		 *
+		 * @param array $params Query params.
+		 *
+		 * @since 1.3.0
+		 */
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: This docblock needs a few changes in order for this to show up in our automated hook docs. See example [here](https://github.com/globeandmail/sophi-for-wordpress/blob/develop/includes/classes/SiteAutomation/Request.php#L77). Main thing to note is the use of `@hook` and the brackets around the params.

File: includes/classes/SiteAutomation/EndPoints.php [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+	 * @param  WP_REST_Request $request The request.
+	 * @return mixed
+	 */
+	public function site_automation_override( $request ) {
+		$current_user = wp_get_current_user();
+
+		if ( $current_user->exists() ) {
+			$user_email = $current_user->user_email;
+		} else {
+			return new \WP_Error( 401, __( 'Unauthorised user, please log in.', 'sophi-wp' ) );
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: I mention this above but probably worth moving this into a `permission_callback` instead of having it here. Also, we only check if a user is currently logged-in/exists, do we need to lock this down to specific user roles?

File: includes/classes/SiteAutomation/EndPoints.php [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+		if ( is_wp_error( $api_token ) ) {
+			return new \WP_Error( 401, __( 'Invalid API token, please try adding correct credentials on the settings page.', 'sophi-wp' ) );
+		}
+
+		// Request parameters.
+		$timeout    = 3;
+		$attributes = $request->get_query_params();
+
+		$api_url = get_sophi_settings( 'sophi_override_url' );
+		$host    = get_sophi_settings( 'host' );
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php:243]: Is there a chance that either of these values wouldn't exist? Or if we have a valid access token (which is required to get to this point) are these values guaranteed to be set?

File: includes/classes/SiteAutomation/EndPoints.php [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+		$api_url = $api_url . 'hosts/' . $host . '/overrides';
+
+		$rule_type        = isset( $attributes['ruleType'] ) ? sanitize_title( $attributes['ruleType'] ) : '';
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: We have this pattern in a few places and this is totally optional to change but we could simplify this code a bit by using the null coalescing operator, so something like
```php
$attributes['ruleType'] ?? ''
```

We do lose the escaping there, but that could be moved to when we use these values in the `$body` below, which is also a little nicer because it's more clear that those values are escaped

File: includes/classes/SiteAutomation/EndPoints.php [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+		$body = array(
+			"articleId"           => $override_post_ID,
+			"expirationHourOfDay" => $override_expiry,
+			"page"                => $page_name,
+			"position"            => 'ban' === $rule_type || 'out' === $rule_type ? '' : $position,
+			"requestedUserName"   => $user_email,
+			"ruleType"            => 'ban' === $rule_type ? 'out' : $rule_type,
+			"widgetName"          => 'ban' === $rule_type ? null : $widget_name,
+		);
+		$body = wp_json_encode( $body );
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: Minor thing but I'd probably remove this line and just add the `wp_json_encode` around the value when we add it to the `body` param

File: includes/classes/SiteAutomation/EndPoints.php [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+			'body'    => $body,
+		];
+
+		/**
+		 * Filters the arguments being passed to the override api auth request.
+		 *
+		 * @param array $args Arguments.
+		 * @param string $api_url Auth API URL.
+		 *
+		 * @since 1.3.0
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: Same comment as the other filter, a few changes needed here for our hookdocs

File: includes/classes/SiteAutomation/EndPoints.php [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+			'sanitize_callback' => 'sanitize_text_field',
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+
+		/**
+		 * Filters the get sophi posts query params.
+		 *
+		 * @param array $params Query params.
+		 *
+		 * @since 1.3.0
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]: Same comment as the other filter, a few changes needed here for our hookdocs

File: includes/functions/settings.php [hunk 16]
```diff
@@ -187,17 +188,67 @@ function fields_setup() {
 		]
 	);
 
+	// Add Auth settings section.
+	add_settings_section(
+			'sophi_api_auth',
```
Comment (dkotter) [inline includes/functions/settings.php]: Appears there's some extra tabs in this file

File: includes/classes/SiteAutomation/Auth.php (hunk 6 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/Auth.php]: Yes we can do that, should we use `md5` while saving the token in the database?

https://github.com/globeandmail/sophi-for-wordpress/blob/98299cf5c0b7657a699d7a11586184fcef74d52c/includes/classes/SiteAutomation/Auth.php#L50

And we then have to decrypt here:

https://github.com/globeandmail/sophi-for-wordpress/blob/98299cf5c0b7657a699d7a11586184fcef74d52c/includes/classes/SiteAutomation/Auth.php#L29

Comment (dkotter) [reply includes/classes/SiteAutomation/Auth.php]: So `md5` is used to hash something (like a password) but is not intended to be used when you need to be able to decrypt a value (which is what we'll need to do in this case).

I'm not an expert when it comes to this and I know there are multiple tools you can use to handle the decryption and multiple ways to deal with the storing of the encryption key. I researched this topic for something else we were working on and I found https://felix-arntz.me/blog/storing-confidential-data-in-wordpress/ to be super helpful here. I'd recommend taking a look at that for a possible approach we could take here.

Comment (jeffpaul) [reply includes/classes/SiteAutomation/Auth.php]: @peterwilsoncc tagging you here in case you've got any advice on potential approach here (unless Felix's notes best apply)?

Comment (peterwilsoncc) [reply includes/classes/SiteAutomation/Auth.php]: Felix's advice seems pretty good to me.

For the key, you can probably use `wp_salt( 'logged_in' )` rather than accessing the php constant direct. This will ensure that you don't end up encrypting with the value `put your unique phrase here`. That will negate to need for a fallback too.

Comment (peterwilsoncc) [reply includes/classes/SiteAutomation/Auth.php]: Oh, something else to keep in mind is that transients are not guaranteed to stick around for the length of time specified in the second parameter of `set_transient()`. In theory they can be deleted immediately (there are plugins that disable them altogether) but more likely they drop out of the cache earlier than expected on sites using a persistent object cache.

tl;dr: if this must stick around then it's better to store as an option `[key => $key, expires => $timestamp ]`. Transients are fine if the time to live is intended as a suggestion only.

File: includes/blocks/sophi-page-list-item/index.css (hunk 5 above)
Comment (faisal-alvi) [reply includes/blocks/sophi-page-list-item/index.css]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/8cd22ada9fe3b5cc7ae8f271e064d16cdf12bf6f

File: includes/classes/SiteAutomation/Auth.php (hunk 6 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/Auth.php]: Thank you for the details every one. Implemented in: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/e911f6a87029c6389d38d1607e8db771558ecda1

@peterwilsoncc even if the transient is deleted before the actual expiry, we create another one when not found. Storing it as an array in the option key would require: getting the array > check for expiry -> use the token, that would take a bit more time IMO.

File: includes/classes/SiteAutomation/EndPoints.php (hunk 8 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/9366ca91567a1e86603d0d324ec6727bce23ab21

File: includes/classes/SiteAutomation/EndPoints.php (hunk 10 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: I think yes, but it depends on what is the specific requirement here. If the Gutenberg block should be accessible to all users who can edit the post, then we are good, but if not, we need to know which roles should be allowed/not-allowed. cc: @jeffpaul

File: includes/classes/SiteAutomation/EndPoints.php (hunk 11 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php:243]: Yes, the values are set, because the admin is not allowed to save the details on the settings page without adding these details.

image

And in order to display a block, admin must have to save all the details on the settings page.

File: includes/classes/SiteAutomation/EndPoints.php (hunk 9 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/d7f3bb5f76704d21c9cf317832c74ef2d0842ca3

File: includes/classes/SiteAutomation/EndPoints.php (hunk 14 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/d7f3bb5f76704d21c9cf317832c74ef2d0842ca3

File: includes/classes/SiteAutomation/EndPoints.php (hunk 15 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/d7f3bb5f76704d21c9cf317832c74ef2d0842ca3

File: includes/classes/SiteAutomation/EndPoints.php (hunk 12 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/86d6081ea65378dfa82a86b0c753ac542a491995

File: includes/classes/SiteAutomation/EndPoints.php (hunk 13 above)
Comment (faisal-alvi) [reply includes/classes/SiteAutomation/EndPoints.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/5c81350a45e6423c30cb9d7111bbdcacc9076af8

File: includes/functions/settings.php (hunk 16 above)
Comment (faisal-alvi) [reply includes/functions/settings.php]: https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/a6e5c9a29c3a800de1252b568bc107ec63904ea4

**Comment by faisal-alvi** [comment]
Please note that #216 is also [fixed](https://github.com/globeandmail/sophi-for-wordpress/pull/333/commits/2ad90892da9785093d2117c3efa3767eb1690cae) by this PR.

File: includes/classes/SiteAutomation/EndPoints.php (hunk 10 above)
Comment (jeffpaul) [reply includes/classes/SiteAutomation/EndPoints.php]: Accessible to all users who can otherwise edit the post seems fine to me

File: includes/classes/SiteAutomation/EndPoints.php [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+	 * Check the permission.
+	 *
+	 * @since 1.3.0
+	 *
+	 * @return bool
+	 */
+	public function site_automation_permission(){
+		$current_user = wp_get_current_user();
+ 
+		if( ! $current_user->exists() ) {
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]:
```suggestion
		if ( ! $current_user->exists() ) { 
```

File: includes/blocks/site-automation-block/edit.js [hunk 18]
```diff
@@ -25,26 +36,234 @@ import { editPropsShape } from './props-shape';
...
+				if (!data?.length) {
+					setMessage({
+						text: __('Posts not found in the site!', 'sophi-wp'),
+						color: 'red',
+					});
+					return false;
+				}
+
+				// eslint-disable-next-line array-callback-return
+				data.map((item) => {
```
Comment (fabiankaegy) [inline includes/blocks/site-automation-block/edit.js]: if you don't care about the returned value you can also use `.forEach` here. It doesn't care about a return value

**Review by dkotter — DISMISSED** [review]

**Comment by jeffpaul** [comment]
@faisal-alvi for the nested block override buttons, please ensure they're labeled `Add`, `Replace`, `Remove`, and `Ban` (so update `Update` to `Replace` and `Trash` to `Ban`).

File: includes/blocks/site-automation-block/register.php [hunk 19]
```diff
@@ -76,7 +81,7 @@ function render_block_callback( $attributes, $content, $block ) {
 				$curated_posts = get_post_meta( $sophi_cached_response->posts[0], 'sophi_site_automation_data', true );
 			}
 		}
-	}
+	}
```
Comment (dkotter) [inline includes/blocks/site-automation-block/register.php]:
```suggestion
	}
```

File: includes/classes/SiteAutomation/EndPoints.php [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+	 * @param int   $post_ID Post ID.
+	 * @param array $rules array of the parent block settings.
+	 *
+	 * @return \stdClass Prepared post data.
+	 */
+	public function get_post_details( $post_ID, $rules ) {
+		$post_data = get_post( $post_ID );
+
+		// If post not found in the site, return the postID only.
+		if( null === $post_data ) {
```
Comment (dkotter) [inline includes/classes/SiteAutomation/EndPoints.php]:
```suggestion
		if ( null === $post_data ) {
```

**Review by dkotter — APPROVED** [review]
